Adding __hash__ and __eq__ to objects

benkiel opened the pull request.

benkiel commented:
This should handle #52. See tests.

benkiel commented:
@behdad / @anthrotype: I know that you guys are both really busy, but if you had the chance to look at this and give it a thumbs up or a "woah, you really got that wrong" it'd be helpful. Understand completely, of course, if you don't have time to review.

justvanrossum commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -24,6 +24,14 @@ def __eq__(self, other):
+        return False
+
+    def __hash__(self):
+        return id(self)
```
I think this is still a problem: two different objects may still compare equal through their `._point` attr, yet will have a different hash.

justvanrossum commented on Lib/fontParts/base/segment.py:
```diff
@@ -16,6 +16,14 @@ def __eq__(self, other):
+
+    def __hash__(self):
+        return id(self)
+
```
Here as well.

justvanrossum commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -24,6 +24,14 @@ def __eq__(self, other):
+    def __ne__(self, other):
+        if hasattr(other, "_point"):
+            return self._point != other._point
+        return False
```
This must be a `True` I think.

justvanrossum commented on Lib/fontParts/base/base.py:
```diff
@@ -52,6 +52,32 @@ def _reprContents(self):
+        """
+        if isinstance(other, self.__class__):
+            return self.naked() is not other.naked()
+        return False
```
This one also.

justvanrossum commented on Lib/fontParts/base/segment.py:
```diff
@@ -16,6 +16,14 @@ def __eq__(self, other):
+    def __ne__(self, other):
+        if hasattr(other, "points"):
+            return self.points != other.points
+        return False
```
And this.

behdad commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -22,15 +22,15 @@ def _setPoint(self, point):
         if hasattr(other, "_point"):
             return self._point == other._point
-        return False
+        return NotImplemented
```
__eq__ shall only return True or False.

benkiel replied:
"A rich comparison method may return the singleton NotImplemented if it does not implement the operation for a given pair of arguments." from the python documentation, so it falls back to NotImplemented when it can't compare two objects that are of unlike types, but I can change that back.

behdad replied:
I thought that only applies to __lt__ etc. But you are right. Python takes care of it.

benkiel replied:
I made that switch when I put in `__ne__` for 2.7, as returning `True` if it can't compare objects didn't make sense.

behdad commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -22,15 +22,15 @@ def _setPoint(self, point):
-            return self._point != other._point
-        return False
+            return not self._point == other._point
+        return NotImplemented
```
Same.

behdad commented on Lib/fontParts/base/segment.py:
```diff
@@ -24,6 +19,19 @@ def _reprContents(self):
+        return NotImplemented
+
+    def __hash__(self):
+        return hash(self.points)
```
I'm confused.

If the aim is to fix https://github.com/robofab-developers/fontParts/issues/52
then you DON'T want this hash function. You just want `id(self)`. Same with __eq__.

What you have now, is not good hashable type, as self.points can change, right?

benkiel replied:
Got it, thank you!

behdad commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -21,16 +21,16 @@ def _setPoint(self, point):
-            return self._point == other._point
-        return NotImplemented
+            return id(self._point) == id(other._point)
+        return False
```
You were correct indeed. You can return NotImplemented here. Python takes care of returning False.

behdad commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -21,16 +21,16 @@ def _setPoint(self, point):
         if hasattr(other, "_point"):
-            return self._point == other._point
-        return NotImplemented
+            return id(self._point) == id(other._point)
```
In that case, looks like "return self._point is other._point" is what you want. You want to return True only if the two _point members point to the same object. Right?

benkiel replied:
Yup, that's the idea, but it's not quite working yet. I'll keep hammering it...

anthrotype commented on Lib/fontParts/base/base.py:
```diff
@@ -52,6 +52,32 @@ def _reprContents(self):
+        Subclasses may override this method.
+        """
+        if isinstance(other, self.__class__):
+            return not self.naked() is other.naked()
```
in python3, `object.__ne__()` automatically delegates to `__eq__()` and inverts the result, unless the latter returns `NotImplemented`. To support both 2 and 3, something like this:

```python
def __ne__(self, other):
    equal = self.__eq__(other)
    return NotImplemented if equal is NotImplemented else not equal
```

anthrotype commented:
sorry, I was away last week. @benkiel I haven't been following the discussion and it's not clear to me what this PR trying to solve. Can you please remind me?

benkiel commented:
@anthrotype is for #52, to make the main objects hashable. I had it so that hash was on the base object, but this may be extreme overkill and making things more complicated than they need to be. I've removed that, updated __ne__ (thank you!) and added has to font and glyph. It may make more sense to add hash to objects as needed, right now those two seem to be the most likely candidates that would actually need it. Sorry this is so messy, will squash when finally merging,

anthrotype commented:
I suggest you have a read at this article to help clarify this issue:
https://hynek.me/articles/hashes-and-equality/

note that if you implement `__eq__` but not `__hash__` then the object is automatically unhashable in python3 (which is a good thing), whereas for python 2 you need to explicitly set `__hash__ = None` in the class scope to mark it as such.

benkiel commented:
@anthrotype thank you, did not know about the 2.7 set to `None`. @justvanrossum sent me that article also, I think that things now abide by it, and it was very helpful

benkiel commented:
Merging this, and closing #52. If more objects need to be hashable, we can add them as needed.
